### PR TITLE
[WIP] cassandra backend

### PR DESCRIPTION
**Pull request by sksumit1**
Cassandra backend for powerdns

### CODE REVIEW

**Comment by sksumit1** [comment]
Need to install the cassandra lib on the box.

pdns_with_cassandra.m4 : Line no. 32, point to the cassandra lib.
For mac it will be dylib, for linux it will be .so file

Building cassandra lib : 
http://datastax.github.io/cpp-driver/topics/building/

**Comment by cmouse** [comment]
Does it support pkg--config?

File: modules/cassandrabackend/cassandrabackend.cc [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+		int reconnect_wait_time = getArgAsNum("reconnect-wait-time");
+		int concurrent_requests_threshold = getArgAsNum("concurrent-requests-threshold");
+		int connect_timeout = getArgAsNum("connect-timeout");
+		int request_timeout = getArgAsNum("request-timeout");
+		int enable_load_balance_round_robin = getArgAsNum("enable-load-balance-round-robin");
+		int enable_token_aware_routing = getArgAsNum("enable-token-aware-routing");
+		int enable_latency_aware_routing = getArgAsNum("enable-latency-aware-routing");
+		int enable_tcp_nodelay = getArgAsNum("enable-tcp-nodelay");
+		int enable_tcp_keepalive = getArgAsNum("enable-tcp-keepalive");
+		all_authoritative_server = getArgAsNum("all-authoritative-server");
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: Can you please explain this feature and why it is required? It seems bit odd place for it.

File: modules/cassandrabackend/cassandrabackend.cc [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+		L << Logger::Info << "[cassandradbmanager] cassandra-queue-size-io " << queue_size_io << " cassandra-queue-size-event " << queue_size_event << endl;
+		L << Logger::Info << "[cassandradbmanager] cassandra-reconnect-wait-time " << reconnect_wait_time << " cassandra-concurrent-requests-threshold " << concurrent_requests_threshold << endl;
+		L << Logger::Info << "[cassandradbmanager] cassandra-connect-timeout " << connect_timeout << " cassandra-request-timeout " << request_timeout << endl;
+		L << Logger::Info << "[cassandradbmanager] cassandra-enable-load-balance-round-robin " << enable_load_balance_round_robin << " cassandra-enable-token-aware-routing " << enable_token_aware_routing << endl;
+		L << Logger::Info << "[cassandradbmanager] cassandra-enable-latency-aware-routing " << enable_latency_aware_routing << endl;
+		L << Logger::Info << "[cassandradbmanager] cassandra-enable-tcp-nodelay " << enable_tcp_nodelay << " cassandra-enable-tcp-keepalive " << enable_tcp_keepalive << endl;
+		L << Logger::Info << "[cassandradbmanager] all-authoritative-server " << all_authoritative_server <<endl;
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: Is it really necessary to log all these on every time this backend is spawned (can happen many times during application lifecycle)

File: modules/cassandrabackend/cassandrabackend.cc [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+		throw;
+	 } catch (...) {
+		throw;
+	 }
+
+  }
+
+  bool list(const string &target, int domain_id, bool include_disabled=false)
+  {
+    return false; // we don't support AXFR
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: Is there reason why AXFR is not supported?

File: modules/cassandrabackend/cassandrabackend.cc [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+#define DEBUGLOG(msg) L<<Logger::Error<<msg
+#else
+#define DEBUGLOG(msg) do {} while(0)
+#endif
+
+class CassandraBackend : public DNSBackend
+{
+
+private:
+  backendrecord* backendRecords = NULL;
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: Perhaps use vector<backendrecord> instead?

File: modules/cassandrabackend/cassandrabackend.cc [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+			 return true;
+		 }
+	}
+	 L << Logger::Info << "[CassandraBackend] SOA record false"<< endl;
+	 clear();
+	 return false;
+
+  }
+
+  void populatedefaults(SOAData &soadata) {
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: You can get default SOA record by leaving these fields untouched, except the db field.

File: modules/cassandrabackend/cassandrabackend.cc [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    fetchdata();
+
+  }
+
+  void fetchdata()
+  {
+	  L << Logger::Info << "[CassandraBackend] Calling cassandradbmanger" << endl;
+	  cassandradbmanager *sc1 = cassandradbmanager::getInstance();
+	  const char* query = "SELECT domain, recordmap, creation_time FROM pdns.domain_lookup_records WHERE domain = ?";
+	  L << Logger::Info << "[CassandraBackend] SELECT domain, recordmap, creation_time FROM pdns.domain_lookup_records WHERE domain = "<<domain<< endl;
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: This seems bit excessive logging, use DLOG() to disable these unless debug compile is made?

Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: Or use query-logging configuration attribute.

File: modules/cassandrabackend/cassandrabackend.cc [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+  {
+	  L << Logger::Info << " Read record Step 1 " << recordIndex << " " << totalSize << endl;
+	  if(this->totalSize == 0 || recordIndex >= this->totalSize) {
+		  L << Logger::Info << " No record False " << recordIndex << " "<< totalSize << endl;
+		  clear();
+		  return false;
+	  }
+	  backendrecord backendRecord = backendRecords[recordIndex];
+	  if((backendRecord.getType() != QType::SOA) && (backendRecord.getType().getName() == queryType || queryType == "ANY")) {
+		  L << Logger::Info << " Read record Step 2 " << endl;
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: Very excessive logging.

File: modules/cassandrabackend/cassandradbmanager.cc [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+        instanceFlag = true;
+        return single;
+    }
+    else
+    {
+        return single;
+    }
+}
+
+void cassandradbmanager::executeQuery(const char* query, struct domainlookuprecords* result1, const char* key, const char* dns_query_type) {
```
Comment (cmouse) [inline modules/cassandrabackend/cassandradbmanager.cc]: pdnssec utility has directBackendCmd which would probably suite here better.

File: modules/cassandrabackend/test/pdns.conf [hunk 9]
```diff
@@ -0,0 +1,552 @@
+# Autogenerated configuration file template
```
Comment (cmouse) [inline modules/cassandrabackend/test/pdns.conf]: You can remove the commented out lines from your config file, they are not necessary for testing.

File: modules/cassandrabackend/test/testresults.txt [hunk 10]
```diff
@@ -0,0 +1,60 @@
+sumit_kumar$ dig ANY cassandra.pdns.com @127.0.0.1
```
Comment (cmouse) [inline modules/cassandrabackend/test/testresults.txt]: The test suite has 'cleandig' utility for producing cleaned out text data.

**Comment by cmouse** [comment]
For testing, see how remotebackend testing has been integrated with the rest of the test suite.

**Comment by cmouse** [comment]
Also, as a note about commit messages, that they should be descriptive, but you can squash them out just before merge. No need to touch them now.

**Comment by sksumit1** [comment]
Hi Aki,

Yes, i will be cutting down on the logging.
I am currently working on the AXFR.
Please give me a few days. I will get back to you with document & the
changes.
You can discard the pull request.

Thanks.
Sumit

**Comment by cmouse** [comment]
You can add [WIP] to the title to indicate that it is not yet finished.

**Comment by cmouse** [comment]
Hi! Why did you close the pull request?

**Comment by sksumit1** [comment]
Hi Aki,

I will be making all the changes suggested by you.
I was thinking of giving a fresh pull request. Should i add to the same?

**Comment by sksumit1** [comment]
I just realized i can use the same pull request.
Not good in Git :(
Reopening

File: modules/cassandrabackend/cassandrabackend.cc (hunk 5 above)
Comment (sksumit1) [inline modules/cassandrabackend/cassandrabackend.cc]: I was thinking of having a configurable SOA, if the user chooses cassandra-all-authoritative-server mode. WIP

**Comment by sksumit1** [comment]
Hi Aki,

Can you review the latest changes please.

Thanks,

File: modules/cassandrabackend/cassandrabackend.cc (hunk 1 above)
Comment (sksumit1) [inline modules/cassandrabackend/cassandrabackend.cc]: The configurations now explained in backend-cassandra.md

**Comment by cmouse** [comment]
you seem to need to rebase your pull.

**Comment by sksumit1** [comment]
Rebased.
However CI is failing. https://travis-ci.org/PowerDNS/pdns/jobs/76422460
that too in "tinydnsbackend". Any idea?

File: pdns/dnsdistdist/m4/dnsdist_lua.m4 [hunk 11]
```diff
@@ -1,39 +0,0 @@
-AC_DEFUN([DNSDIST_LUA],[
```
Comment (cmouse) [inline pdns/dnsdistdist/m4/dnsdist_lua.m4]: why are you removing this file?

File: modules/cassandrabackend/backendutil.cc [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+		int size = 0;
+		if(result != NULL) {
+			std::map<std::string, records> records_list = result->recordTypeResultArrayMap;
+			for(it_type_string_record iterator = records_list.begin(); iterator != records_list.end(); iterator++) {
+				records record = iterator->second;
+				size+=record.recordMap.size();
+			}
+		}
+		result->size = size;
+		backendrecord* output = new backendrecord[size];
```
Comment (cmouse) [inline modules/cassandrabackend/backendutil.cc]: vector would serve better here.

**Comment by cmouse** [comment]
Please remove Makefile and Makefile.in from your pull =)

**Comment by cmouse** [comment]
I don't think we should have the all-authoritative-backend in one backend. If you need/want this feature, I think you should create separate pull request to implement this feature in core. Or you can use script backend for this.

File: modules/cassandrabackend/test/testresults.txt (hunk 10 above)
Comment (sksumit1) [inline modules/cassandrabackend/test/testresults.txt]: Will make use of it

**Comment by cmouse** [comment]
the tinydns failure is curious, it seems to be caused by dual definition of uint64_t in boost and std namespace. Which is strange since your patch does not seem to touch tinydns.

**Comment by sksumit1** [comment]
Hi Aki,

CI is faling because of tinydns.

The CI is passing -tinydns.
https://api.travis-ci.org/jobs/76587876/log.txt?deansi=true

On including tinydns i get the below errors.

[0mIn file included from tinydnsbackend.cc:4:
[1m./tinydnsbackend.hh:76:2: [0m[0;1;31merror: [0m[1mreference to
'uint64_t' is ambiguous[0m
        uint64_t d_taiepoch;
[0;1;32m        ^
[0m[1m/usr/include/stdint.h:55:27: [0m[0;1;30mnote: [0mcandidate
found by name lookup is 'uint64_t'[0m
typedef unsigned long int       uint64_t;
[0;1;32m                                ^
[0m[1m/usr/include/boost/cstdint.hpp:311:42: [0m[0;1;30mnote:
[0mcandidate found by name lookup is
      'boost::uint64_t'[0m
     typedef  ::boost::ulong_long_type   uint64_t;
[0;1;32m                                         ^
[0m[1mtinydnsbackend.cc:263:4: [0m[0;1;31merror: [0m[1mreference
to 'uint64_t' is ambiguous[0m
                        uint64_t timestamp = pr.get32BitInt();
[0;1;32m                        ^
[0m[1m/usr/include/stdint.h:55:27: [0m[0;1;30mnote: [0mcandidate
found by name lookup is 'uint64_t'[0m
typedef unsigned long int       uint64_t;
[0;1;32m                                ^
[0m[1m/usr/include/boost/cstdint.hpp:311:42: [0m[0;1;30mnote:
[0mcandidate found by name lookup is
      'boost::uint64_t'[0m
     typedef  ::boost::ulong_long_type   uint64_t;
[0;1;32m                                         ^
[0m[1mtinydnsbackend.cc:264:4: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'timestamp'[0m
                        timestamp <<= 32;
[0;1;32m                        ^
[0m[1mtinydnsbackend.cc:265:4: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'timestamp'[0m
                        timestamp += pr.get32BitInt();
[0;1;32m                        ^
[0m[1mtinydnsbackend.cc:266:7: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'timestamp'[0m
                        if(timestamp) {
[0;1;32m                           ^
[0m[1mtinydnsbackend.cc:267:5: [0m[0;1;31merror: [0m[1mreference
to 'uint64_t' is ambiguous[0m
                                uint64_t now = d_taiepoch + time(NULL);
[0;1;32m                                ^
[0m[1m/usr/include/stdint.h:55:27: [0m[0;1;30mnote: [0mcandidate
found by name lookup is 'uint64_t'[0m
typedef unsigned long int       uint64_t;
[0;1;32m                                ^
[0m[1m/usr/include/boost/cstdint.hpp:311:42: [0m[0;1;30mnote:
[0mcandidate found by name lookup is
      'boost::uint64_t'[0m
     typedef  ::boost::ulong_long_type   uint64_t;
[0;1;32m                                         ^
[0m[1mtinydnsbackend.cc:269:10: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'timestamp'[0m
                                        if (timestamp < now) {
[0;1;32m                                            ^
[0m[1mtinydnsbackend.cc:269:22: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'now'[0m
                                        if (timestamp < now) {
[0;1;32m                                                        ^
[0m[1mtinydnsbackend.cc:272:15: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'timestamp'[0m
                                        rr.ttl = timestamp - now;
[0;1;32m                                                 ^
[0m[1mtinydnsbackend.cc:272:27: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'now'[0m
                                        rr.ttl = timestamp - now;
[0;1;32m                                                             ^
[0m[1mtinydnsbackend.cc:273:16: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'now'[0m
                                } else if (now <= timestamp) {
[0;1;32m                                           ^
[0m[1mtinydnsbackend.cc:273:23: [0m[0;1;31merror: [0m[1muse of
undeclared identifier 'timestamp'[0m
                                } else if (now <= timestamp) {
[0;1;32m                                                  ^
[0m2 warnings and 12 errors generated.
make[3]: *** [tinydnsbackend.lo] Error 1


I tried to reproduce at my local setup, but couldn't.
Any idea?

WIP: Unit tests

Thanks,
Sumit

**Comment by sksumit1** [comment]
WIP : Unit tests

**Comment by sksumit1** [comment]
FYI, Boost version on my laptop is 1.55.
Travis is using 1.46

**Comment by cmouse** [comment]
You seem to update libc6 and libc6-dev to a more recent version in travis preparations. This would seem to be the cause of the trouble.

**Comment by sksumit1** [comment]
I really am having a tough time understanding how to write regression tests for the backend. Is there any document which explains its working & how to write them.

**Comment by sksumit1** [comment]
Just a update incase you are watching this pull.
I am using cassandra 2.1.8.
Apparently travis has some problems with cassandra 2.1.x (https://github.com/travis-ci/travis-ci/issues/4626)
I have tried lots of workarounds, but apparently the build is failing with no reason.
I will wait and watch for travis to come up with cassandra 2.1.x support.

File: docs/markdown/authoritative/backend-cassandra.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+|:--|:--|
+|Native|Yes|
+|Master|No|
+|Slave|No|
+|Superslave|No|
+|DNSSEC|No|
+|Module name|cassandra|
+|Launch name|cassandra|
+
+This is a backend powered by a columnar NoSql database [Apache Cassandra](http://cassandra.apache.org/), ritten for the 3.4 release & above, with easily configurable cassandra cluster parameters.
```
Comment (simi) [inline docs/markdown/authoritative/backend-cassandra.md]: s/ritten/written/g

Comment (sksumit1) [inline docs/markdown/authoritative/backend-cassandra.md]: Thanks for pointing out. Fixed.

File: .travis.yml [hunk 14]
```diff
@@ -33,9 +35,10 @@ before_script:
      libopendbx1-dev
      libopendbx1-sqlite3
      libp11-kit-dev
+     libssl-dev
      libtolua-dev
      libtool
-     links
+     links
```
Comment (cmouse) [inline .travis.yml]: whitespace nit

File: .travis.yml [hunk 15]
```diff
@@ -16,15 +18,15 @@ before_script:
...
      bc
      bind9utils
+     cmake
      daemontools
      dnsutils
      faketime
      geoip-database
      ldnsutils
-     libboost-all-dev
+     libboost1.55-all-dev
```
Comment (cmouse) [inline .travis.yml]: why is this changed?

File: modules/cassandrabackend/cassandrabackend.cc [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+						L << Logger::Info << "[CassandraBackend] SOA record found" << endl;
+					}
+					soadata.db = this;
+					//Data format :: domain_id nameserver	hostname    serial refresh retry expiry default_ttl
+					std::string data = backendRecord.getRecord();
+					boost::char_separator<char> delimiter("#");
+					boost::tokenizer<boost::char_separator<char> > tokens(data,
+							delimiter);
+					int rec_index = 0;
+					BOOST_FOREACH(std::string const& token, tokens) {
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: Maybe instead of this just validate the size of the tokens vector and access with index directly?

File: modules/cassandrabackend/cassandrabackend.cc [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+		try {
+			queryType = type.getName();
+			domain = qname.toStringNoDot();
+			recordIndex = 0;
+			totalSize = 0;
+			if(::arg().mustDo("query-logging")) {
+				L << Logger::Info << "[CassandraBackend] Recieved lookup query for " << queryType << " " << domain << " " << endl;
+			}
+			fetchdata();
+		} catch (const std::exception &exc) {
```
Comment (cmouse) [inline modules/cassandrabackend/cassandrabackend.cc]: You should probably throw a PDNSException here

File: modules/tinydnsbackend/tinydnsbackend.cc [hunk 18]
```diff
@@ -260,11 +260,11 @@ bool TinyDNSBackend::get(DNSResourceRecord &rr)
 			rr.auth = true;
 
 			rr.ttl = pr.get32BitInt();
-			uint64_t timestamp = pr.get32BitInt();
```
Comment (cmouse) [inline modules/tinydnsbackend/tinydnsbackend.cc]: this should be a separate pull request, and you can use std::uint64_t here.

File: modules/tinydnsbackend/tinydnsbackend.hh [hunk 19]
```diff
@@ -73,7 +73,7 @@ private:
 	typedef TDI_t::index<tag_domainid>::type TDIById_t;
 
 	//data member variables
-	uint64_t d_taiepoch;
+	unsigned long long d_taiepoch;
```
Comment (cmouse) [inline modules/tinydnsbackend/tinydnsbackend.hh]: this should be a separate pull request, and you can use std::uint64_t here.

File: pdns/dnsdistdist/m4/dnsdist_lua.m4 [hunk 20]
```diff
@@ -36,4 +36,3 @@ AC_DEFUN([DNSDIST_LUA],[
   ])
   AM_CONDITIONAL([LUA], [test "x$with_lua" = "xyes"])
 ])
-
```
Comment (cmouse) [inline pdns/dnsdistdist/m4/dnsdist_lua.m4]: unnecessary whitespace change

File: modules/cassandrabackend/cassandrabackend.cc (hunk 17 above)
Comment (sksumit1) [inline modules/cassandrabackend/cassandrabackend.cc]: throwing PDNSException now

File: modules/tinydnsbackend/tinydnsbackend.hh (hunk 19 above)
Comment (sksumit1) [inline modules/tinydnsbackend/tinydnsbackend.hh]: This can't be a separate pull request.
Cassandra needs upgrade of libc6 and libc6-dev, this causes failure to tinydns due to namespace conflict between std::uint64_t and boost::uint64_t

File: modules/tinydnsbackend/tinydnsbackend.cc (hunk 18 above)
Comment (sksumit1) [inline modules/tinydnsbackend/tinydnsbackend.cc]: This can't be a separate pull request.
Cassandra needs upgrade of libc6 and libc6-dev, this causes failure to tinydns due to namespace conflict between std::uint64_t and boost::uint64_t

Comment (cmouse) [inline modules/tinydnsbackend/tinydnsbackend.cc]: Yes it can be. We are able to merge future pulls before this.

**Comment by Habbie** [comment]
@sksumit1 what is the status of this? Does it work? Is it complete? We'd love to have a Cassandra backend!

**Comment by sksumit1** [comment]
Hi Peter,

I last worked on it in Sept 2015. At that time due to travis limitations, I
had a hard time installing cassandra 2.2
Since cassandra is a resource intensive db, if i started it before the test
suite, OOM killer would kill it, as a result cassandra's tests will fail. I
am now starting cassandrajust before executing its test suite. Cassandra's
uptime was not predictable as a result test suites were failing.
Can you confirm the current version of cassandra on travis. (When i last
checked it was cassandra 1.2)
I had been following the travis upgrade on this issue
https://github.com/travis-ci/travis-ci/issues/3952, but there is still no
resolution.
I can redo my work if the travis supports min cassandra 2.1.5.

Thanks,
Sumit

**Comment by Habbie** [comment]
Closing this due to inactivity. Please open a new PR if this work is revived.
